core/services/ocr2/plugins/ocr2keeper/evm21/core/payload.go

```
@@ -0,0 +1,86 @@
+package core
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	ocr2keepers "github.com/smartcontractkit/ocr2keepers/pkg"
+)
+
+// UpkeepTriggerID returns the identifier using the given upkeepID and trigger.
+// It follows the same logic as the contract, but performs it locally.
+func UpkeepTriggerID(id *big.Int, trigger []byte) string {
+	idBytes := id.Bytes()
+
+	combined := append(idBytes, trigger...)
+
+	triggerIDBytes := crypto.Keccak256(combined)
+	triggerID := hex.EncodeToString(triggerIDBytes)
+
+	return triggerID
+}
+
+// UpkeepWorkID returns the identifier using the given upkeepID and trigger extension(tx hash and log index).
+func UpkeepWorkID(id *big.Int, trigger ocr2keepers.Trigger) (string, error) {
+	extensionBytes, err := json.Marshal(trigger.Extension)
+	if err != nil {
+		return "", err
+	}
+
+	// TODO (auto-4314): Ensure it works with conditionals and add unit tests
+	combined := fmt.Sprintf("%s%s", id, extensionBytes)
+	hash := crypto.Keccak256([]byte(combined))
+	return hex.EncodeToString(hash[:]), nil
+}
+
+func NewUpkeepPayload(uid *big.Int, tp int, trigger ocr2keepers.Trigger, checkData []byte) (ocr2keepers.UpkeepPayload, error) {
+	// construct payload
+	p := ocr2keepers.UpkeepPayload{
+		Upkeep: ocr2keepers.ConfiguredUpkeep{
+			ID:     ocr2keepers.UpkeepIdentifier(uid.Bytes()),
+			Type:   tp,
+			Config: struct{}{}, // empty struct by default
+		},
+		Trigger:   trigger,
+		CheckData: checkData,
+	}
+	// set work id based on upkeep id and trigger
+	p.WorkID, _ = UpkeepWorkID(uid, trigger)
```

handle ignored error

---

done